### ISSUE TITLE
Fixing Pad

### ISSUE THREAD

**Issue by skewballfox** [secretsauceai/mfcc-rust#4]
Something I noticed last night was that the we need to rework the last argument of Pad to be a string match. in the issue where we got this pad function, they were padding with a constant value, and in the python code, they were using two different padding styles(that I noticed): edge and symmetric. from the [docs](https://numpy.org/doc/stable/reference/generated/numpy.pad.html?highlight=pad#numpy.pad)

> ‘edge’
>  Pads with the edge values of array.

[numpy source](https://github.com/numpy/numpy/blob/v1.22.0/numpy/lib/arraypad.py#L821)

numpy example
```python
>>> a = [1, 2, 3, 4, 5]
>>> np.pad(a, (2, 3), 'edge')
array([1, 1, 1, ..., 5, 5, 5])
```
> ‘symmetric’
>  Pads with the reflection of the vector mirrored along the edge of the array.

[numpy source](https://github.com/numpy/numpy/blob/v1.22.0/numpy/lib/arraypad.py#L844)
numpy example 
```python
>>> a = [1, 2, 3, 4, 5]
>>> np.pad(a, (2, 3), 'symmetric')
array([2, 1, 1, 2, 3, 4, 5, 5, 4, 3])
```

**Comment by sheosi** [comment]
This is effectively an enum. While old Python used strings instead, in Rust we can translate that as an enum.
```
pub enum PadMode {
    Edge,
    Symmetric
}
```

And then just reference `PadMode::Edge` or `PadMode::Symmetric`.

**Comment by skewballfox** [comment]
yeah, I figured an Enum(with match) would be a better fit for this. the
main problem with fixing pad is the current implementation (which I
shamelessly ripped from [ndarry#823](
https://github.com/rust-ndarray/ndarray/issues/823) in order to have some
starting point) is assuming a constant padding value, whereas numpy pad
drastically changes the operation(and the necessary iterations) depending
on the padding type.

we can either implement multiple pad functions(one per pad type), or make a
function where the last arg is a enum.

**Comment by skewballfox** [comment]
looking over the arguments, the second one(`pad_width`) is an array-like where each element-pair are the padding sizes at the beginning and end of a given axis. here's an example of what happens to a 2d 4x5 array of 20 elements:
![image](https://user-images.githubusercontent.com/29311431/151683521-4f205376-5782-4e24-8bc6-dd5dd8d7f05d.png)

the original array starts at the 2nd row, 4th column, and ends at the 5th row, 9th column. note what happens to the created 2d blocks at the corners of the original 2dmatrix: they are filled with the corner value, or reflected along the extended axis(not sure how to describe that)

**Comment by skewballfox** [comment]
a really clean implementation we may want to borrow from is the [pad function from ndarray-ndimage](https://github.com/imeka/ndarray-ndimage/blob/master/src/pad.rs), though that is for 3 dimensional arrays

**Comment by skewballfox** [comment]
the implementation I'm working on currently spits out this when fed a (4,5) d array and a pad width of `[[20,20],[0,0]]`. here's the list of things that need to be fixed just to fit our specific use case(where one axis's padding is always 0):
- [ ] the top and bottom sections need to be filled
- [ ] the array needs to be flipped on each other tile
- [ ] partial sections at the border need to be supported
- [ ] edge type needs to be handled
```
[[0.0, 0.0, 0.0, 0.0, 0.0],
 [0.0, 0.0, 0.0, 0.0, 0.0],
 [0.0, 0.0, 0.0, 0.0, 0.0],
 [0.0, 0.0, 0.0, 0.0, 0.0],
 [0.0, 1.0, 2.0, 3.0, 4.0],
 [5.0, 6.0, 7.0, 8.0, 9.0],
 [10.0, 11.0, 12.0, 13.0, 14.0],
 [15.0, 16.0, 17.0, 18.0, 19.0],
 [0.0, 1.0, 2.0, 3.0, 4.0],
 [5.0, 6.0, 7.0, 8.0, 9.0],
 [10.0, 11.0, 12.0, 13.0, 14.0],
 [15.0, 16.0, 17.0, 18.0, 19.0],
 [0.0, 1.0, 2.0, 3.0, 4.0],
 [5.0, 6.0, 7.0, 8.0, 9.0],
 [10.0, 11.0, 12.0, 13.0, 14.0],
 [15.0, 16.0, 17.0, 18.0, 19.0],
 [0.0, 1.0, 2.0, 3.0, 4.0],
 [5.0, 6.0, 7.0, 8.0, 9.0],
 [10.0, 11.0, 12.0, 13.0, 14.0],
 [15.0, 16.0, 17.0, 18.0, 19.0],
 [0.0, 1.0, 2.0, 3.0, 4.0],
 [5.0, 6.0, 7.0, 8.0, 9.0],
 [10.0, 11.0, 12.0, 13.0, 14.0],
 [15.0, 16.0, 17.0, 18.0, 19.0],
 [0.0, 1.0, 2.0, 3.0, 4.0],
 [5.0, 6.0, 7.0, 8.0, 9.0],
 [10.0, 11.0, 12.0, 13.0, 14.0],
 [15.0, 16.0, 17.0, 18.0, 19.0],
 [0.0, 1.0, 2.0, 3.0, 4.0],
 [5.0, 6.0, 7.0, 8.0, 9.0],
 [10.0, 11.0, 12.0, 13.0, 14.0],
 [15.0, 16.0, 17.0, 18.0, 19.0],
 [0.0, 1.0, 2.0, 3.0, 4.0],
 [5.0, 6.0, 7.0, 8.0, 9.0],
 [10.0, 11.0, 12.0, 13.0, 14.0],
 [15.0, 16.0, 17.0, 18.0, 19.0],
 [0.0, 1.0, 2.0, 3.0, 4.0],
 [5.0, 6.0, 7.0, 8.0, 9.0],
 [10.0, 11.0, 12.0, 13.0, 14.0],
 [15.0, 16.0, 17.0, 18.0, 19.0],
 [0.0, 0.0, 0.0, 0.0, 0.0],
 [0.0, 0.0, 0.0, 0.0, 0.0],
 [0.0, 0.0, 0.0, 0.0, 0.0],
 [0.0, 0.0, 0.0, 0.0, 0.0]], shape=[44, 5], strides=[5, 1], layout=Cc (0x5), const ndim=2

```

**Comment by skewballfox** [comment]
the first draft of pad has been implemented in #8
